### PR TITLE
Explicitly filter on repo name like we do on branch_name

### PR DESCRIPTION
**Pull request by boakley**
Maybe this fixes performance issues on production when filtering on repo_name, but I'm able to test that assertion locally. 

TL;DR: filtering on `branch_name` works, `repo_name` times out.  I had implemented them differently; this change implements `repo_name` like `branch_name` as an experiment.

Specifically, these queries worked:

* https://sfdo-metaci.herokuapp.com/api/robot/?range=today
*https://sfdo-metaci.herokuapp.com/api/robot/?range=today&branch_name=feature/robot__winter21fix

A third variation where I filter on `repo_name` instead of `branch_name` would time out:

* https://sfdo-metaci.herokuapp.com/api/robot/?range=today&repo_name=NPSP

I don't have a good grasp on how django rest framework filters and when they are applied. However, in looking at the code I noticed that I handled `branch_name` and `repo_name` differently.  With `branch_name` I was explicitly pulling that parameter from the URL and applying it to a query of builds prior to doing a query for test results. With `repo_name` I was _not_ doing this, instead relying on the filterset which I'm guessing is applied after all of the test results were queried.

### CODE REVIEW

**Comment by prescod** [comment]
I could be wrong but my guess is that this won't make things any better or worse. :(

Have you seen the Django debug toolbar? It can let you see your queries.

**Comment by davisagli** [comment]
The Django debug toolbar is not going to be very useful at debugging this view (which returns a csv file) or at identifying which queries are slow in production (where the toolbar isn't enabled).

But I agree that we should try to actually understand what is slow, not just make stabs in the dark.

From the Diagnose tab of the Heroku postgres addon, I identified this query which is very slow and looks related:
```
SELECT COUNT(*) AS "__count"
FROM "testresults_testresult"
INNER JOIN "build_buildflow"
    ON ("testresults_testresult"."build_flow_id" = "build_buildflow"."id")
INNER JOIN "testresults_testmethod"
    ON ("testresults_testresult"."method_id" = "testresults_testmethod"."id")
INNER JOIN "testresults_testclass"
    ON ("testresults_testmethod"."testclass_id" = "testresults_testclass"."id")
INNER JOIN "build_build"
    ON ("build_buildflow"."build_id" = "build_build"."id")
INNER JOIN "repository_repository"
    ON ("build_build"."repo_id" = "repository_repository"."id")
WHERE ("testresults_testresult"."build_flow_id" IN 
    (SELECT V0."id"
    FROM "build_buildflow" V0
    INNER JOIN "build_build" V1
        ON (V0."build_id" = V1."id")
    WHERE (V1."planrepo_id" IN 
        (SELECT "plan_planrepository"."id"
        FROM "plan_planrepository")
                AND (V0."time_end" AT TIME ZONE ?)::date >= ?::date
                AND (V0."time_end" AT TIME ZONE ?)::date < ?::date))
                AND "testresults_testclass"."test_type" = ?
                AND UPPER("repository_repository"."name"::text) = UPPER(?))
```

(I'm not sure why a COUNT query is needed, but nevermind, we can still analyze it to understand why it's slow.)

I filled in real values for the ? parameters and ran it as an EXPLAIN query in the postgres shell:
```
EXPLAIN SELECT COUNT(*) AS "__count"
  FROM "testresults_testresult"
 INNER JOIN "build_buildflow"
    ON ("testresults_testresult"."build_flow_id" = "build_buildflow"."id")
 INNER JOIN "testresults_testmethod"
    ON ("testresults_testresult"."method_id" = "testresults_testmethod"."id")
 INNER JOIN "testresults_testclass"
    ON ("testresults_testmethod"."testclass_id" = "testresults_testclass"."id")
 INNER JOIN "build_build"
    ON ("build_buildflow"."build_id" = "build_build"."id")
 INNER JOIN "repository_repository"
    ON ("build_build"."repo_id" = "repository_repository"."id")
 WHERE ("testresults_testresult"."build_flow_id" IN (
        SELECT V0."id"
          FROM "build_buildflow" V0
         INNER JOIN "build_build" V1
            ON (V0."build_id" = V1."id")
         WHERE (V1."planrepo_id" IN (SELECT "plan_planrepository"."id" FROM "plan_planrepository") AND (V0."time_end" AT TIME ZONE 'UTC')::date >= '2020-09-01'::date AND (V0."time_end" AT TIME ZONE 'UTC')::date < '2020-10-01'::date)
       )
   AND "testresults_testclass"."test_type" = 'Robot'
   AND UPPER("repository_repository"."name"::text) = UPPER('NPSP')
 );
```

Result:
```
                                                                                     QUERY PLAN                                                                                     
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=2709382.13..2709382.13 rows=1 width=8)
   ->  Hash Semi Join  (cost=71601.10..2709367.67 rows=28903 width=0)
         Hash Cond: (testresults_testresult.build_flow_id = v0.id)
         ->  Gather  (cost=12042.28..2649522.07 rows=368727 width=8)
               Workers Planned: 1
               ->  Hash Join  (cost=11042.28..2611649.37 rows=216898 width=8)
                     Hash Cond: (testresults_testresult.method_id = testresults_testmethod.id)
                     ->  Nested Loop  (cost=10649.03..2608971.68 rows=2178333 width=12)
                           ->  Hash Join  (cost=10648.92..63857.60 rows=5014 width=4)
                                 Hash Cond: (build_buildflow.build_id = build_build.id)
                                 ->  Parallel Seq Scan on build_buildflow  (cost=0.00..53016.92 rows=235639 width=8)
                                 ->  Hash  (cost=10625.51..10625.51 rows=6689 width=4)
                                       ->  Nested Loop  (cost=0.08..10625.51 rows=6689 width=4)
                                             ->  Seq Scan on repository_repository  (cost=0.00..1.19 rows=1 width=4)
                                                   Filter: (upper((name)::text) = 'NPSP'::text)
                                             ->  Index Scan using build_build_9a8c79bf on build_build  (cost=0.08..10603.81 rows=6835 width=8)
                                                   Index Cond: (repo_id = repository_repository.id)
                           ->  Index Scan using testresults_testresult_d88dce3b on testresults_testresult  (cost=0.11..487.17 rows=6811 width=8)
                                 Index Cond: (build_flow_id = build_buildflow.id)
                     ->  Hash  (cost=384.67..384.67 rows=2452 width=4)
                           ->  Hash Join  (cost=42.03..384.67 rows=2452 width=4)
                                 Hash Cond: (testresults_testmethod.testclass_id = testresults_testclass.id)
                                 ->  Seq Scan on testresults_testmethod  (cost=0.00..316.88 rows=24626 width=8)
                                 ->  Hash  (cost=40.81..40.81 rows=348 width=4)
                                       ->  Bitmap Heap Scan on testresults_testclass  (cost=4.60..40.81 rows=348 width=4)
                                             Recheck Cond: ((test_type)::text = 'Robot'::text)
                                             ->  Bitmap Index Scan on testresults_testclass_test_type_86bff2a8_like  (cost=0.00..4.58 rows=348 width=0)
                                                   Index Cond: ((test_type)::text = 'Robot'::text)
         ->  Hash  (cost=59551.80..59551.80 rows=2003 width=4)
               ->  Gather  (cost=1007.68..59551.80 rows=2003 width=4)
                     Workers Planned: 1
                     ->  Hash Semi Join  (cost=7.68..58351.50 rows=1178 width=4)
                           Hash Cond: (v1.planrepo_id = plan_planrepository.id)
                           ->  Nested Loop  (cost=0.08..58339.54 rows=1178 width=8)
                                 ->  Parallel Seq Scan on build_buildflow v0  (cost=0.00..53723.83 rows=1178 width=8)
                                       Filter: (((timezone('UTC'::text, time_end))::date >= '2020-09-01'::date) AND ((timezone('UTC'::text, time_end))::date < '2020-10-01'::date))
                                 ->  Index Scan using build_build_pkey on build_build v1  (cost=0.08..3.92 rows=1 width=8)
                                       Index Cond: (id = v0.build_id)
                           ->  Hash  (cost=5.66..5.66 rows=553 width=4)
                                 ->  Seq Scan on plan_planrepository  (cost=0.00..5.66 rows=553 width=4)
```

I'm not very experienced in analyzing query plans myself, but some things I notice are:
1. There are intermediate stages with up to 368727 rows
2. There are 5 Seq Scans which means reading the entire table one row at a time because there's no index. It looks like we would probably benefit from adding indexes on the non-PK fields that are being used for the joins, as well as buildflow's time_start & time_end.

**Comment by boakley** [comment]
I don't know if that count query is related or not. I have no experience matching heroku logs to django code, but I know I didn't intentionally ask for a count. Maybe that's added by django. I also have zero experience trying to understand the output of explain. This is just not something I have any experience it at all. :-(

**Comment by davisagli** [comment]
@boakley I'm convinced that the query is related; it's doing the joins and filtering that are implied by the Django queryset. I would guess that the count might be coming from DRF's pagination feature, but that's just a hunch. I would guess you can find information on how to read a postgres query plan if you look.

**Comment by prescod** [comment]
Stumbled no this old thread. Yes the Count is almost certainly related to pagination. If the Count itself is the problem, it can be turned off or replaced (e.g. with cursor pagination which does not need a count)

https://www.django-rest-framework.org/api-guide/pagination/

**Comment by boakley** [comment]
I'm closing this since it's very old, and we should be able to meet the reporting needs of QE via our integration with Tableau.
